File: android/app/src/test/java/io/nexy/android/ui/chat/ChatTurnReducerTest.kt
package io.nexy.android.ui.chat

import io.nexy.android.data.model.WsEvent
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ChatTurnReducerTest {
    @Test
    fun startsTurnAndAccumulatesTextInSequenceOrder() {
        var state = emptyChatTurnState("conv-1")
        state = reduceChatTurn(state, turnEvent("turn_started", 1))
        state = reduceChatTurn(state, turnEvent("assistant_text_delta", 2, """"chunk":"Hello""""))
        state = reduceChatTurn(state, turnEvent("assistant_text_delta", 3, """"chunk":" world""""))

        assertEquals(ChatTurnStatus.Streaming, state.status)
        assertEquals("Hello world", state.text)
        assertEquals(3L, state.lastSequence)
    }

    @Test
    fun ignoresStaleEventsAndEventsForOtherConversations() {
        var state = emptyChatTurnState("conv-1")
        state = reduceChatTurn(state, turnEvent("turn_started", 1))
        state = reduceChatTurn(state, turnEvent("assistant_text_delta", 3, """"chunk":"new""""))
        state = reduceChatTurn(state, turnEvent("assistant_text_delta", 2, """"chunk":" stale""""))
        state = reduceChatTurn(state, turnEvent("assistant_text_delta", 4, """"chunk":" other"""", conversationId = "conv-2"))

        assertEquals("new", state.text)
        assertEquals(3L, state.lastSequence)
    }

    @Test
    fun queuesThinkingDoneBeforeMatchingDeltaAndReplaysIt() {
        var state = emptyChatTurnState("conv-1")
        state = reduceChatTurn(state, turnEvent("turn_started", 1))
        state = reduceChatTurn(state, turnEvent("thinking_done", 2, """"blockId":"reasoning-1""""))

        assertTrue("reasoning-1" in state.pendingThinkingEnds)
        assertTrue(state.thinkingBlocks.isEmpty())

        state = reduceChatTurn(
            state,
            turnEvent("thinking_delta", 3, """"blockId":"reasoning-1","chunk":"Planned steps""""),
        )

        assertEquals(1, state.thinkingBlocks.size)
        assertEquals("Planned steps", state.thinkingBlocks.single().content)
        assertTrue(state.thinkingBlocks.single().done)
        assertTrue(state.pendingThinkingEnds.isEmpty())
    }

    @Test
    fun marksThinkingDoneOnCompletionAndFailure() {
        var completed = emptyChatTurnState("conv-1")
        completed = reduceChatTurn(completed, turnEvent("turn_started", 1))
        completed = reduceChatTurn(completed, turnEvent("thinking_delta", 2, """"blockId":"reasoning-1","chunk":"Still thinking""""))
        completed = reduceChatTurn(completed, turnEvent("turn_completed", 3))

        assertEquals(ChatTurnStatus.Completed, completed.status)
        assertTrue(completed.thinkingBlocks.single().done)

        var failed = emptyChatTurnState("conv-1")
        failed = reduceChatTurn(failed, turnEvent("turn_started", 1))
        failed = reduceChatTurn(failed, turnEvent("thinking_delta", 2, """"blockId":"reasoning-1","chunk":"Still thinking""""))
        failed = reduceChatTurn(
            failed,
            turnEvent("turn_failed", 3, """"errorType":"api","message":"Provider failed","retryable":true"""),
        )

        assertEquals(ChatTurnStatus.Failed, failed.status)
        assertEquals("Provider failed", failed.error?.message)
        assertTrue(failed.thinkingBlocks.single().done)
    }

    @Test
    fun recordsToolModelActivityAndCost() {
        var state = emptyChatTurnState("conv-1")
        state = reduceChatTurn(state, turnEvent("turn_started", 1))
        state = reduceChatTurn(state, turnEvent("model_changed", 2, """"model":"gpt-5-mini""""))
        state = reduceChatTurn(
            state,
            turnEvent("activity_changed", 3, """"state":"tool","label":"Running browser_snapshot","toolName":"browser_snapshot","serverName":"Browser""""),
        )
        state = reduceChatTurn(
            state,
            turnEvent("tool_finished", 4, """"toolName":"browser_snapshot","serverName":"Browser","args":{"tab":"active"},"result":"Snapshot captured","success":true"""),
        )
        state = reduceChatTurn(
            state,
            turnEvent("cost_updated", 5, """"inputTokens":100,"outputTokens":25,"totalCostUsd":0.01"""),
        )

        assertEquals("gpt-5-mini", state.model)
        assertEquals("Running browser_snapshot", state.activity?.label)
        assertEquals("browser_snapshot", state.toolCalls.single().toolName)
        assertEquals("""{"tab":"active"}""", state.toolCalls.single().argsJson)
        assertEquals(100, state.cost?.inputTokens)
        assertEquals(25, state.cost?.outputTokens)
        assertEquals(0.01, state.cost?.totalCostUsd ?: 0.0, 0.0001)
        assertNull(state.error)
    }

    @Test
    fun updatesDuplicateToolFinishedEventsWithTheSameId() {
        var state = emptyChatTurnState("conv-1")
        state = reduceChatTurn(state, turnEvent("turn_started", 1))
        state = reduceChatTurn(
            state,
            turnEvent("tool_finished", 2, """"id":"tool-1","toolName":"read_file","result":"old","success":true"""),
        )
        state = reduceChatTurn(
            state,
            turnEvent("tool_finished", 3, """"id":"tool-1","toolName":"read_file","result":"new","success":false"""),
        )

        assertEquals(1, state.toolCalls.size)
        assertEquals("tool-1", state.toolCalls.single().id)
        assertEquals("new", state.toolCalls.single().result)
        assertEquals(false, state.toolCalls.single().success)
    }

    private fun turnEvent(
        type: String,
        sequence: Long,
        payload: String = "",
        conversationId: String = "conv-1",
        turnId: String = "turn-1",
    ): WsEvent.ChatTurnEvent {
        val payloadJson = buildString {
            append("""{"type":"$type","conversationId":"$conversationId","turnId":"$turnId","sequence":$sequence,"timestamp":${1000 + sequence}""")
            if (payload.isNotBlank()) append(",").append(payload)
            append("}")
        }
        return WsEvent.ChatTurnEvent(
            conversationId = conversationId,
            turnId = turnId,
            sequence = sequence,
            type = type,
            timestamp = 1000 + sequence,
            payloadJson = payloadJson,
        )
    }
}

File: src/renderer/__tests__/chat-turn-reducer.test.ts
import { describe, expect, it } from 'vitest'
import type { ChatTurnEvent } from '../../shared/chat-turn-types'
import { chatTurnReducer, createEmptyChatTurnState } from '../hooks/chat-turn-reducer'

function event(overrides: Partial<ChatTurnEvent> & { type: ChatTurnEvent['type']; sequence: number }): ChatTurnEvent {
  return {
    conversationId: 'conv-1',
    turnId: 'turn-1',
    timestamp: 1000 + overrides.sequence,
    ...overrides,
  } as ChatTurnEvent
}

describe('chatTurnReducer', () => {
  it('starts a turn and accumulates assistant text deltas in sequence order', () => {
    let state = createEmptyChatTurnState('conv-1')
    state = chatTurnReducer(state, event({ type: 'turn_started', sequence: 1 }))
    state = chatTurnReducer(state, event({ type: 'assistant_text_delta', sequence: 2, chunk: 'Hello' }))
    state = chatTurnReducer(state, event({ type: 'assistant_text_delta', sequence: 3, chunk: ' world' }))

    expect(state.status).toBe('streaming')
    expect(state.text).toBe('Hello world')
    expect(state.lastSequence).toBe(3)
  })

  it('ignores stale sequence numbers and events for other conversations', () => {
    let state = createEmptyChatTurnState('conv-1')
    state = chatTurnReducer(state, event({ type: 'turn_started', sequence: 1 }))
    state = chatTurnReducer(state, event({ type: 'assistant_text_delta', sequence: 3, chunk: 'new' }))
    state = chatTurnReducer(state, event({ type: 'assistant_text_delta', sequence: 2, chunk: ' stale' }))
    state = chatTurnReducer(state, event({
      type: 'assistant_text_delta',
      sequence: 4,
      conversationId: 'conv-2',
      chunk: ' other',
    }))

    expect(state.text).toBe('new')
    expect(state.lastSequence).toBe(3)
  })

  it('queues thinking_done before the matching delta and replays it', () => {
    let state = createEmptyChatTurnState('conv-1')
    state = chatTurnReducer(state, event({ type: 'turn_started', sequence: 1 }))
    state = chatTurnReducer(state, event({ type: 'thinking_done', sequence: 2, blockId: 'reasoning-1' }))
    expect(state.thinkingBlocks.get('reasoning-1')).toBeUndefined()
    expect(state.pendingThinkingEnds.has('reasoning-1')).toBe(true)

    state = chatTurnReducer(state, event({
      type: 'thinking_delta',
      sequence: 3,
      blockId: 'reasoning-1',
      chunk: 'Planned steps',
    }))

    expect(state.thinkingBlocks.get('reasoning-1')).toEqual({
      blockId: 'reasoning-1',
      content: 'Planned steps',
      done: true,
      firstSeenSequence: 3,
    })
    expect(state.pendingThinkingEnds.size).toBe(0)
  })

  it('marks all thinking blocks done on completion and failure', () => {
    let completed = createEmptyChatTurnState('conv-1')
    completed = chatTurnReducer(completed, event({ type: 'turn_started', sequence: 1 }))
    completed = chatTurnReducer(completed, event({
      type: 'thinking_delta',
      sequence: 2,
      blockId: 'reasoning-1',
      chunk: 'Still thinking',
    }))
    completed = chatTurnReducer(completed, event({ type: 'turn_completed', sequence: 3 }))

    expect(completed.status).toBe('completed')
    expect(completed.thinkingBlocks.get('reasoning-1')?.done).toBe(true)

    let failed = createEmptyChatTurnState('conv-1')
    failed = chatTurnReducer(failed, event({ type: 'turn_started', sequence: 1 }))
    failed = chatTurnReducer(failed, event({
      type: 'thinking_delta',
      sequence: 2,
      blockId: 'reasoning-1',
      chunk: 'Still thinking',
    }))
    failed = chatTurnReducer(failed, event({
      type: 'turn_failed',
      sequence: 3,
      errorType: 'api',
      message: 'Provider failed',
      retryable: true,
    }))

    expect(failed.status).toBe('failed')
    expect(failed.error?.message).toBe('Provider failed')
    expect(failed.thinkingBlocks.get('reasoning-1')?.done).toBe(true)
  })

  it('records tool calls, model, activity, and cost', () => {
    let state = createEmptyChatTurnState('conv-1')
    state = chatTurnReducer(state, event({ type: 'turn_started', sequence: 1 }))
    state = chatTurnReducer(state, event({ type: 'model_changed', sequence: 2, model: 'gpt-5-mini' }))
    state = chatTurnReducer(state, event({
      type: 'activity_changed',
      sequence: 3,
      state: 'tool',
      label: 'Running browser_snapshot',
      toolName: 'browser_snapshot',
      serverName: 'Browser',
    }))
    state = chatTurnReducer(state, event({
      type: 'tool_finished',
      sequence: 4,
      toolName: 'browser_snapshot',
      serverName: 'Browser',
      args: { tab: 'active' },
      result: 'Snapshot captured',
      success: true,
    }))
    state = chatTurnReducer(state, event({
      type: 'cost_updated',
      sequence: 5,
      inputTokens: 100,
      outputTokens: 25,
      totalCostUsd: 0.01,
    }))

    expect(state.model).toBe('gpt-5-mini')
    expect(state.activity).toEqual({
      state: 'tool',
      label: 'Running browser_snapshot',
      toolName: 'browser_snapshot',
      serverName: 'Browser',
    })
    expect(state.toolCalls).toEqual([{
      id: undefined,
      toolName: 'browser_snapshot',
      serverName: 'Browser',
      args: { tab: 'active' },
      result: 'Snapshot captured',
      success: true,
      resultImages: undefined,
      inProgress: false,
      firstSeenSequence: 4,
    }])
    expect(state.cost).toEqual({
      inputTokens: 100,
      outputTokens: 25,
      totalCostUsd: 0.01,
    })
  })

  it('inserts an in-progress placeholder on tool_started and resolves it on tool_finished', () => {
    let state = createEmptyChatTurnState('conv-1')
    state = chatTurnReducer(state, event({ type: 'turn_started', sequence: 1 }))
    state = chatTurnReducer(state, event({
      type: 'tool_started',
      sequence: 2,
      id: 'call-1',
      name: 'read_file',
      input: { path: 'a.ts' },
      serverName: 'claude-cli',
    }))

    expect(state.toolCalls).toEqual([{
      id: 'call-1',
      toolName: 'read_file',
      serverName: 'claude-cli',
      args: { path: 'a.ts' },
      result: '',
      success: true,
      inProgress: true,
      firstSeenSequence: 2,
    }])

    state = chatTurnReducer(state, event({
      type: 'tool_finished',
      sequence: 3,
      id: 'call-1',
      toolName: 'read_file',
      serverName: 'claude-cli',
      result: 'file contents',
      success: true,
    }))

    expect(state.toolCalls).toEqual([{
      id: 'call-1',
      toolName: 'read_file',
      serverName: 'claude-cli',
      args: undefined,
      result: 'file contents',
      success: true,
      resultImages: undefined,
      inProgress: false,
      firstSeenSequence: 2,
    }])
  })

  it('updates duplicate tool_finished events with the same id', () => {
    let state = createEmptyChatTurnState('conv-1')
    state = chatTurnReducer(state, event({ type: 'turn_started', sequence: 1 }))
    state = chatTurnReducer(state, event({
      type: 'tool_finished',
      sequence: 2,
      id: 'tool-1',
      toolName: 'read_file',
      result: 'old',
      success: true,
    }))
    state = chatTurnReducer(state, event({
      type: 'tool_finished',
      sequence: 3,
      id: 'tool-1',
      toolName: 'read_file',
      result: 'new',
      success: false,
    }))

    expect(state.toolCalls).toEqual([{
      id: 'tool-1',
      toolName: 'read_file',
      serverName: undefined,
      args: undefined,
      result: 'new',
      success: false,
      resultImages: undefined,
      inProgress: false,
      firstSeenSequence: 2,
    }])
  })
})
